spi/zynq7000-spi/libzynq7000-spi.c

```diff
@@ -0,0 +1,518 @@
+			{ SPI_MODE0, SPI_SCLK_MAX }, /* SPI0 SS1 configuration */
+			{ SPI_MODE0, SPI_SCLK_MAX }, /* SPI0 SS2 configuration */
+		},
+	},
+	{
+		.id = 1,             /* SPI1 */
+		.irq = 81,           /* SPI1 IRQ */
+		.paddr = 0xe0007000, /* SPI1 base physical address */
+		.pins = {
+			{ pctl_mio_pin_10, MIO_MOSI }, /* SPI1 MOSI */
+			{ pctl_mio_pin_11, MIO_MISO }, /* SPI1 MISO */
+			{ pctl_mio_pin_12, MIO_SCLK }, /* SPI1 SCLK */
+			{ pctl_mio_pin_13, MIO_SS },   /* SPI1 SS0 */
+			{ pctl_mio_pin_14, MIO_SS },   /* SPI1 SS1 */
+			{ pctl_mio_pin_15, MIO_SS },   /* SPI1 SS2 */
+		},
+		.slaves = {
+			{ SPI_MODE0, SPI_SCLK_MAX }, /* SPI1 SS0 configuration */
+			{ SPI_MODE0, SPI_SCLK_MAX }, /* SPI1 SS1 configuration */
+			{ SPI_MODE0, SPI_SCLK_MAX }, /* SPI1 SS2 configuration */
+		},
+	},
+};
+
+
+static spi_t *spi_get(unsigned int ndev)
+{
+	if (ndev >= sizeof(devs) / sizeof(devs[0])) {
+		return NULL;
+	}
+
+	return &devs[ndev];
+}
+
+
+static int spi_isr(unsigned int n, void *arg)
+{
+	spi_t *spi = arg;
+	unsigned int status;
+
+	status = *(spi->base + sr);
+
+	/* Clear all interrupts */
+	*(spi->base + sr) = status;
+
+	/* TX FIFO empty */
+	if (status & (1 << 2)) {
+		*(spi->base + idr) = (1 << 2);
+		return spi->cond;
+	}
+
+	return 0;
+}
+
+
+#if 0
+static unsigned char spi_getMode(spi_t *spi)
+{
+	unsigned char mode = 0;
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+
+	if (cfg & (1 << 2)) {
+		mode |= SPI_CPHA;
+	}
+
+	if (cfg & (1 << 1)) {
+		mode |= SPI_CPOL;
+	}
+
+	return mode;
+}
+#endif
+
+
+static void spi_setMode(spi_t *spi, unsigned char mode)
+{
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+	cfg &= ~((1 << 2) | (1 << 1));
+
+	if (mode & SPI_CPHA) {
+		cfg |= (1 << 2);
+	}
+
+	if (mode & SPI_CPOL) {
+		cfg |= (1 << 1);
+	}
+
+	*(spi->base + cr) = cfg;
+}
+
+
+int spi_mode(unsigned int ndev, unsigned int ss, unsigned char mode)
+{
+	spi_t *spi;
+
+	spi = spi_get(ndev);
+	if (spi == NULL) {
+		return -ENODEV;
+	}
+
+	if (ss >= sizeof(spi->slaves) / sizeof(spi->slaves[0])) {
+		return -EINVAL;
+	}
+
+	spi->slaves[ss].mode = mode;
+
+	return EOK;
+}
+
+
+#if 0
+static unsigned int spi_getSpeed(spi_t *spi)
+{
+	unsigned int cfg;
+	unsigned char divisor;
+
+	cfg = *(spi->base + cr);
+	divisor = (cfg >> 3) & 7;
+
+	return SPI_SCLK(divisor);
+}
+#endif
+
+
+static void spi_setSpeed(spi_t *spi, unsigned int speed)
+{
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+	cfg &= ~(7 << 3);
+	cfg |= (__builtin_ffs(SPI_CLK_REF / speed) - 2) << 3;
+
+	*(spi->base + cr) = cfg;
+}
+
+
+int spi_speed(unsigned int ndev, unsigned int ss, unsigned int speed)
+{
+	spi_t *spi;
+	unsigned char divisor;
+
+	spi = spi_get(ndev);
+	if (spi == NULL) {
+		return -ENODEV;
+	}
+
+	if (ss >= sizeof(spi->slaves) / sizeof(spi->slaves[0])) {
+		return -EINVAL;
+	}
+
+	for (divisor = SPI_CLK_MIN_DIV; (divisor < SPI_CLK_MAX_DIV) && (SPI_SCLK(divisor) > speed); divisor++)
+		;
+
+	spi->slaves[ss].speed = SPI_SCLK(divisor);
+
+	return spi->slaves[ss].speed;
```

Comment:
The returned value meant to be the actual SCLK clock set in the controller (it may not be exact `speed` passed to the function due to limited available dividers).
However, I've changed the SPI interface so that there is a separate function to get the SCLK speed - `spi_getSpeed()`. In effect the returned value in question got changed to `EOK`.